src/server/multi_command_squasher.cc

```diff
@@ -26,18 +26,24 @@ template <typename F> void IterateKeys(CmdArgList args, KeyIndex keys, F&& f) {
 
 MultiCommandSquasher::MultiCommandSquasher(absl::Span<StoredCmd> cmds, ConnectionContext* cntx,
                                            bool error_abort)
-    : cmds_{cmds}, cntx_{cntx}, base_cid_{cntx->transaction->GetCId()}, error_abort_{error_abort} {
+    : cmds_{cmds}, cntx_{cntx}, base_cid_{nullptr}, error_abort_{error_abort} {
   auto mode = cntx->transaction->GetMultiMode();
-  track_keys_ = mode == Transaction::NON_ATOMIC;
+  base_cid_ = mode == Transaction::NON_ATOMIC ? nullptr : cntx->transaction->GetCId();
 }
 
 MultiCommandSquasher::ShardExecInfo& MultiCommandSquasher::PrepareShardInfo(ShardId sid) {
   if (sharded_.empty())
     sharded_.resize(shard_set->size());
 
   auto& sinfo = sharded_[sid];
-  if (!sinfo.local_tx)
-    sinfo.local_tx = new Transaction{cntx_->transaction};
+  if (!sinfo.local_tx) {
+    if (base_cid_) {
+      sinfo.local_tx = new Transaction{cntx_->transaction};
+    } else {
+      sinfo.local_tx = new Transaction{cntx_->transaction->GetCId(), sid};
+      sinfo.local_tx->StartMultiNonAtomic();
+    }
+  }
```

Comment:
I feel like this needs a comment, by I've already commented twice on the difference of non-atomic/atomic

Reply:
you can add a comment where you comment about it :)